# Capstone Data Science Project - The Battle of Neighborhoods (Both the weeks combined)

## Description

### Discussion of the Background

As indicated by Bloomberg News, the London Housing Market is stuck. It is presently confronting various headwinds, including the possibility of higher assessments and an admonition from the Bank of England that U.K. home estimations could fall as much as 30 percent in case of a scattered exit from the European Union. All the more explicitly, four neglected splits recommend that the London market might be fit as a fiddle than many acknowledge: concealed value falls, record-low deals, homebuilder departure and assessment climbs tending to abroad purchasers of homes in England and Wales.

### Discussion of the Problem

In this situation, it is dire to receive AI devices so as to help homebuyers customers in London to settle on insightful and powerful choices. Therefore, the business issue we are presently presenting is: how might we offer help to homebuyers demographic in to buy a reasonable land in London in this unsure monetary and money related situation? To take care of this business issue, we are going to group London neighborhoods so as to suggest settings and the present normal cost of land where homebuyers can make a land speculation. We will prescribe beneficial settings as indicated by civilities and fundamental offices encompassing such scenes for example grade schools, secondary schools, emergency clinics and markets.

### Data used

Information on London properties and the relative cost paid information were separated from the HM Land Registry (http://landregistry.data.gov.uk/). The accompanying fields contain the location information remembered for Price Paid Data: Postcode; PAON Primary Addressable Object Name. Commonly the house number or name; SAON Secondary Addressable Object Name. In the event that there is a sub-working, for instance, the structure is separated into pads, there will be a SAON; Street; Locality; Town/City; District; County. To investigate and target prescribed areas across various scenes as indicated by the nearness of conveniences and basic offices, we will get to information through FourSquare API interface and mastermind them as a dataframe for representation. By blending information on London properties and the relative cost paid information from the HM Land Registry and information on courtesies and basic offices encompassing such properties from FourSquare API interface, we will have the option to suggest productive land investments.

### Method Description

1. We will collect the data
2. Explore the data and Try to Understand it.
3. Prepare the data and process the data.
4. Data Modelling

### Step 1: Let's Collect the data

Importing the required libraries and downloading the data 

In [56]:
import pandas as pd
import numpy as np
import datetime as dt
import os 
import json
!conda install -c conda-forge geopy --yes
#Conversion Latitude and Longitude
from geopy.geocoders import Nominatim 
# Used to handle requests
import requests
# Transformation of json to pandas
from pandas.io.json import json_normalize 
# For plotting
import matplotlib.cm as cm
import matplotlib.colors as colors
!conda install -c conda-forge folium=0.5.0 --yes
import folium

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.



In [57]:
# Reading the Data from the source
df_ppd = pd.read_csv("http://prod2.publicdata.landregistry.gov.uk.s3-website-eu-west-1.amazonaws.com/pp-2018.csv")

### Step 2: Exploring the Data and Trying to Understand it

Displaying the first 5 rows using head

In [58]:
df_ppd.head(5)

{79A74E21-D11E-1289-E053-6B04A8C01627}  770000  2018-09-25 00:00  SK7 1AR  \
0  {79A74E21-D11F-1289-E053-6B04A8C01627}  253500  2018-09-24 00:00   M6 8GQ   
1  {79A74E21-D120-1289-E053-6B04A8C01627}  231950  2018-09-28 00:00  WA3 2UE   
2  {79A74E21-D121-1289-E053-6B04A8C01627}  112500  2018-08-29 00:00  OL6 6RJ   
3  {79A74E21-D122-1289-E053-6B04A8C01627}  184995  2018-06-15 00:00  M46 0TW   
4  {79A74E21-D123-1289-E053-6B04A8C01627}  214995  2018-09-28 00:00  M28 3XS   

   D  N  F    5 Unnamed: 8          OAK MEADOW  BRAMHALL          STOCKPORT  \
0  D  N  F    1        NaN      RIVINGTON ROAD       NaN            SALFORD   
1  D  Y  F   35        NaN     STONEACRE CLOSE    LOWTON         WARRINGTON   
2  S  N  F  102        NaN    THORNFIELD GROVE       NaN  ASHTON-UNDER-LYNE   
3  S  Y  F   37        NaN  THREADNEEDLE PLACE  ATHERTON         MANCHESTER   
4  D  Y  L    9        NaN      MARPLE GARDENS   WORSLEY         MANCHESTER   

  STOCKPORT.1  GREATER MANCHESTER  A A.1  
0     SALFORD  GREATER MANCHESTER  A   A  
1       WIGAN  GREATER MANCHESTER  A   A  
2    TAMESIDE  GREATER MANCHESTER  A   A  
3       WIGAN  GREATER MANCHESTER  A   A  
4     SALFORD  GREATER MANCHESTER  A   A

In [59]:
df_ppd.shape

(1029749, 16)

The dataset has 1029749 rows and 16 columns

### Step 3:  Prepare the Data and Process the Data

At this stage, we set up our dataset for the demonstrating procedure, deciding on the most appropriate AI calculation for our extension. As needs be, we play out the accompanying advances: Rename the segment names,Organization the date section, Sort information by date of offer, Select information just for the city of London, Make a rundown of road names in London, Figure the road insightful normal cost of the property, Peruse the road savvy facilitates into an information outline, disposing of repeating word London from singular names, Join the information to discover the directions of areas which fit into customer's financial plan, Plot suggested areas on London map alongside current market costs.

In [60]:
df_ppd.columns = ['TUID', 'Price', 'Date_Transfer', 'Postcode', 'Prop_Type', 'Old_New', 'Duration', 'PAON', \
                  'SAON', 'Street', 'Locality', 'Town_City', 'District', 'County', 'PPD_Cat_Type', 'Record_Status']

df_ppd['Date_Transfer'] = df_ppd['Date_Transfer'].apply(pd.to_datetime) # Formatting the date coln
df_ppd.drop(df_ppd[df_ppd.Date_Transfer.dt.year < 2016].index, inplace=True) # Clearing off every old exchange which were done before 2016
df_ppd.sort_values(by=['Date_Transfer'],ascending=[False],inplace=True) # Sorting by sale date

In [61]:
df_ppd_london = df_ppd.query("Town_City == 'LONDON'")
streets = df_ppd_london['Street'].unique().tolist() # List of name of the streets in London
df_grp_price = df_ppd_london.groupby(['Street'])['Price'].mean().reset_index()
df_grp_price.columns = ['Street', 'Avg_Price']
df_affordable = df_grp_price.query("(Avg_Price >= 2200000) & (Avg_Price <= 2500000)") # Locations which fits the budget between the two limits

In [62]:
df_affordable # Affordable street with average price

Street     Avg_Price
196            ALBION SQUARE  2.450000e+06
390              ANHALT ROAD  2.435000e+06
405          ANSDELL TERRACE  2.250000e+06
422          APPLEGARTH ROAD  2.400000e+06
855          BARONSMEAD ROAD  2.375000e+06
981           BEAUCLERC ROAD  2.480000e+06
1102         BELVEDERE DRIVE  2.340000e+06
1215       BICKENHALL STREET  2.208500e+06
1253       BIRCHLANDS AVENUE  2.217000e+06
1553          BRAMPTON GROVE  2.456875e+06
1632       BRIARDALE GARDENS  2.397132e+06
1797                BROOKWAY  2.400000e+06
1914            BURBAGE ROAD  2.445000e+06
1980               BURY WALK  2.492500e+06
2068         CALLCOTT STREET  2.375000e+06
2129       CAMPDEN HILL ROAD  2.379653e+06
2136            CAMPION ROAD  2.461000e+06
2158           CANNING PLACE  2.425000e+06
2225           CARLISLE ROAD  2.200000e+06
2230         CARLTON GARDENS  2.483500e+06
2242           CARLYLE COURT  2.300000e+06
2406          CHALCOT SQUARE  2.286679e+06
2484            CHARLES LANE  2.414000e+06
2562        CHELSEA CRESCENT  2.495000e+06
2606     CHESTER CLOSE NORTH  2.450000e+06
2638            CHEYNE COURT  2.250000e+06
2641              CHEYNE ROW  2.410000e+06
2686           CHISWICK MALL  2.287500e+06
2761               CITY ROAD  2.468340e+06
2808        CLARENDON STREET  2.250000e+06
...                      ...           ...
10935   RUSSELL GARDENS MEWS  2.300000e+06
11182         SETTLES STREET  2.487500e+06
11257         SHELDON AVENUE  2.349542e+06
11491          SOUTH END ROW  2.470000e+06
11568    SOUTHWOOD LAWN ROAD  2.350000e+06
11571         SOVEREIGN PARK  2.500000e+06
11788  ST MARGARETS CRESCENT  2.216500e+06
11824       ST OSWALDS PLACE  2.250000e+06
11844       ST PETERS SQUARE  2.468730e+06
11881       STAFFORD TERRACE  2.355000e+06
12221       SUTHERLAND PLACE  2.456000e+06
12283          SYDNEY STREET  2.240833e+06
12425            THAMES BANK  2.400000e+06
12487            THE HEXAGON  2.335000e+06
12766        TREDEGAR SQUARE  2.436666e+06
12828         TRINITY STREET  2.317500e+06
12978   UPPER HAMPSTEAD WALK  2.500000e+06
13240        WALPOLE GARDENS  2.303500e+06
13242         WALPOLE STREET  2.242500e+06
13316         WARWICK SQUARE  2.432273e+06
13415            WELBECK WAY  2.267000e+06
13427      WELLESLEY TERRACE  2.410000e+06
13438      WELLINGTON STREET  2.293155e+06
13570     WESTMORELAND PLACE  2.300000e+06
13677       WHITFIELD STREET  2.451000e+06
13722         WILFRED STREET  2.410538e+06
13748     WILLOW BRIDGE ROAD  2.425000e+06
13768          WILSON STREET  2.257500e+06
13796        WINCHENDON ROAD  2.350000e+06
13833           WINGATE ROAD  2.206400e+06

[162 rows x 2 columns]

In [63]:
import pandas as pd
import numpy as np
import datetime as DT
import hmac
from geopy.geocoders import Nominatim
from geopy.distance import vincenty
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [64]:
for index, item in df_affordable.iterrows():
    print(f"index: {index}")
    print(f"item: {item}")
    print(f"item.Street only: {item.Street}")

index: 196
item: Street       ALBION SQUARE
Avg_Price         2.45e+06
Name: 196, dtype: object
item.Street only: ALBION SQUARE
index: 390
item: Street       ANHALT ROAD
Avg_Price      2.435e+06
Name: 390, dtype: object
item.Street only: ANHALT ROAD
index: 405
item: Street       ANSDELL TERRACE
Avg_Price           2.25e+06
Name: 405, dtype: object
item.Street only: ANSDELL TERRACE
index: 422
item: Street       APPLEGARTH ROAD
Avg_Price            2.4e+06
Name: 422, dtype: object
item.Street only: APPLEGARTH ROAD
index: 855
item: Street       BARONSMEAD ROAD
Avg_Price          2.375e+06
Name: 855, dtype: object
item.Street only: BARONSMEAD ROAD
index: 981
item: Street       BEAUCLERC ROAD
Avg_Price          2.48e+06
Name: 981, dtype: object
item.Street only: BEAUCLERC ROAD
index: 1102
item: Street       BELVEDERE DRIVE
Avg_Price           2.34e+06
Name: 1102, dtype: object
item.Street only: BELVEDERE DRIVE
index: 1215
item: Street       BICKENHALL STREET
Avg_Price           2.2085e+06
N

In [65]:
geolocator = Nominatim()

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:1: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  if __name__ == '__main__':


In [66]:
df_affordable['city_coord'] = df_affordable['Street'].apply(geolocator.geocode).apply(lambda x: (x.latitude, x.longitude))
df_affordable

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Street     Avg_Price  \
196            ALBION SQUARE  2.450000e+06   
390              ANHALT ROAD  2.435000e+06   
405          ANSDELL TERRACE  2.250000e+06   
422          APPLEGARTH ROAD  2.400000e+06   
855          BARONSMEAD ROAD  2.375000e+06   
981           BEAUCLERC ROAD  2.480000e+06   
1102         BELVEDERE DRIVE  2.340000e+06   
1215       BICKENHALL STREET  2.208500e+06   
1253       BIRCHLANDS AVENUE  2.217000e+06   
1553          BRAMPTON GROVE  2.456875e+06   
1632       BRIARDALE GARDENS  2.397132e+06   
1797                BROOKWAY  2.400000e+06   
1914            BURBAGE ROAD  2.445000e+06   
1980               BURY WALK  2.492500e+06   
2068         CALLCOTT STREET  2.375000e+06   
2129       CAMPDEN HILL ROAD  2.379653e+06   
2136            CAMPION ROAD  2.461000e+06   
2158           CANNING PLACE  2.425000e+06   
2225           CARLISLE ROAD  2.200000e+06   
2230         CARLTON GARDENS  2.483500e+06   
2242           CARLYLE COURT  2.300000e+06   
2406          CHALCOT SQUARE  2.286679e+06   
2484            CHARLES LANE  2.414000e+06   
2562        CHELSEA CRESCENT  2.495000e+06   
2606     CHESTER CLOSE NORTH  2.450000e+06   
2638            CHEYNE COURT  2.250000e+06   
2641              CHEYNE ROW  2.410000e+06   
2686           CHISWICK MALL  2.287500e+06   
2761               CITY ROAD  2.468340e+06   
2808        CLARENDON STREET  2.250000e+06   
...                      ...           ...   
10935   RUSSELL GARDENS MEWS  2.300000e+06   
11182         SETTLES STREET  2.487500e+06   
11257         SHELDON AVENUE  2.349542e+06   
11491          SOUTH END ROW  2.470000e+06   
11568    SOUTHWOOD LAWN ROAD  2.350000e+06   
11571         SOVEREIGN PARK  2.500000e+06   
11788  ST MARGARETS CRESCENT  2.216500e+06   
11824       ST OSWALDS PLACE  2.250000e+06   
11844       ST PETERS SQUARE  2.468730e+06   
11881       STAFFORD TERRACE  2.355000e+06   
12221       SUTHERLAND PLACE  2.456000e+06   
12283          SYDNEY STREET  2.240833e+06   
12425            THAMES BANK  2.400000e+06   
12487            THE HEXAGON  2.335000e+06   
12766        TREDEGAR SQUARE  2.436666e+06   
12828         TRINITY STREET  2.317500e+06   
12978   UPPER HAMPSTEAD WALK  2.500000e+06   
13240        WALPOLE GARDENS  2.303500e+06   
13242         WALPOLE STREET  2.242500e+06   
13316         WARWICK SQUARE  2.432273e+06   
13415            WELBECK WAY  2.267000e+06   
13427      WELLESLEY TERRACE  2.410000e+06   
13438      WELLINGTON STREET  2.293155e+06   
13570     WESTMORELAND PLACE  2.300000e+06   
13677       WHITFIELD STREET  2.451000e+06   
13722         WILFRED STREET  2.410538e+06   
13748     WILLOW BRIDGE ROAD  2.425000e+06   
13768          WILSON STREET  2.257500e+06   
13796        WINCHENDON ROAD  2.350000e+06   
13833           WINGATE ROAD  2.206400e+06   

                                      city_coord  
196           (-41.27375755, 173.28939323910353)  
390                     (51.4803164, -0.1668011)  
405                     (51.4998899, -0.1891027)  
422                     (53.7486539, -0.3266704)  
855                      (51.4773147, -0.239457)  
981                    (30.2114523, -81.6179807)  
1102                   (38.0728178, -78.4587964)  
1215                    (51.5212014, -0.1589082)  
1253                    (51.4483941, -0.1604676)  
1553                    (51.5899607, -0.3185249)  
1632                    (51.5601748, -0.1954305)  
1797   (45.432184899999996, -122.80281166115779)  
1914                    (52.5385071, -1.3536736)  
1980                    (52.1455294, -0.4235933)  
2068                    (51.5083499, -0.1983276)  
2129                    (51.5081106, -0.1996673)  
2136                      (52.681078, 0.9655987)  
2158                     (52.6410267, -1.135223)  
2225                    (43.3994382, -79.973215)  
2230           (-37.80194335, 144.9719701710172)  
2242    (32.972700950000004, -97.17339170977195)  
2406                    (51.5410876, -0.1559279

In [67]:
df_affordable[['Latitude', 'Longitude']] = df_affordable['city_coord'].apply(pd.Series)
df_affordable

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/frame.py:3391: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[k1] = value[k2]


Street     Avg_Price  \
196            ALBION SQUARE  2.450000e+06   
390              ANHALT ROAD  2.435000e+06   
405          ANSDELL TERRACE  2.250000e+06   
422          APPLEGARTH ROAD  2.400000e+06   
855          BARONSMEAD ROAD  2.375000e+06   
981           BEAUCLERC ROAD  2.480000e+06   
1102         BELVEDERE DRIVE  2.340000e+06   
1215       BICKENHALL STREET  2.208500e+06   
1253       BIRCHLANDS AVENUE  2.217000e+06   
1553          BRAMPTON GROVE  2.456875e+06   
1632       BRIARDALE GARDENS  2.397132e+06   
1797                BROOKWAY  2.400000e+06   
1914            BURBAGE ROAD  2.445000e+06   
1980               BURY WALK  2.492500e+06   
2068         CALLCOTT STREET  2.375000e+06   
2129       CAMPDEN HILL ROAD  2.379653e+06   
2136            CAMPION ROAD  2.461000e+06   
2158           CANNING PLACE  2.425000e+06   
2225           CARLISLE ROAD  2.200000e+06   
2230         CARLTON GARDENS  2.483500e+06   
2242           CARLYLE COURT  2.300000e+06   
2406          CHALCOT SQUARE  2.286679e+06   
2484            CHARLES LANE  2.414000e+06   
2562        CHELSEA CRESCENT  2.495000e+06   
2606     CHESTER CLOSE NORTH  2.450000e+06   
2638            CHEYNE COURT  2.250000e+06   
2641              CHEYNE ROW  2.410000e+06   
2686           CHISWICK MALL  2.287500e+06   
2761               CITY ROAD  2.468340e+06   
2808        CLARENDON STREET  2.250000e+06   
...                      ...           ...   
10935   RUSSELL GARDENS MEWS  2.300000e+06   
11182         SETTLES STREET  2.487500e+06   
11257         SHELDON AVENUE  2.349542e+06   
11491          SOUTH END ROW  2.470000e+06   
11568    SOUTHWOOD LAWN ROAD  2.350000e+06   
11571         SOVEREIGN PARK  2.500000e+06   
11788  ST MARGARETS CRESCENT  2.216500e+06   
11824       ST OSWALDS PLACE  2.250000e+06   
11844       ST PETERS SQUARE  2.468730e+06   
11881       STAFFORD TERRACE  2.355000e+06   
12221       SUTHERLAND PLACE  2.456000e+06   
12283          SYDNEY STREET  2.240833e+06   
12425            THAMES BANK  2.400000e+06   
12487            THE HEXAGON  2.335000e+06   
12766        TREDEGAR SQUARE  2.436666e+06   
12828         TRINITY STREET  2.317500e+06   
12978   UPPER HAMPSTEAD WALK  2.500000e+06   
13240        WALPOLE GARDENS  2.303500e+06   
13242         WALPOLE STREET  2.242500e+06   
13316         WARWICK SQUARE  2.432273e+06   
13415            WELBECK WAY  2.267000e+06   
13427      WELLESLEY TERRACE  2.410000e+06   
13438      WELLINGTON STREET  2.293155e+06   
13570     WESTMORELAND PLACE  2.300000e+06   
13677       WHITFIELD STREET  2.451000e+06   
13722         WILFRED STREET  2.410538e+06   
13748     WILLOW BRIDGE ROAD  2.425000e+06   
13768          WILSON STREET  2.257500e+06   
13796        WINCHENDON ROAD  2.350000e+06   
13833           WINGATE ROAD  2.206400e+06   

                                      city_coord   Latitude   Longitude  
196           (-41.27375755, 173.28939323910353) -41.273758  173.289393  
390                     (51.4803164, -0.1668011)  51.480316   -0.166801  
405                     (51.4998899, -0.1891027)  51.499890   -0.189103  
422                     (53.7486539, -0.3266704)  53.748654   -0.326670  
855                      (51.4773147, -0.239457)  51.477315   -0.239457  
981                    (30.2114523, -81.6179807)  30.211452  -81.617981  
1102                   (38.0728178, -78.4587964)  38.072818  -78.458796  
1215                    (51.5212014, -0.1589082)  51.521201   -0.158908  
1253                    (51.4483941, -0.1604676)  51.448394   -0.160468  
1553                    (51.5899607, -0.3185249)  51.589961   -0.318525  
1632                    (51.5601748, -0.1954305)  51.560175   -0.195431  
1797   (45.432184899999996, -122.80281166115779)  45.432185 -122.802812  
1914                    (52.5385071, -1.3536736)  52.538507   -1.353674  
1980                    (52.1455294, -0.4235933)  52.145529   -0.423593  
2068                    (51.5083499, -0.1983276)  51.508350

In [68]:
df = df_affordable.drop(columns=['city_coord'])
df

Street     Avg_Price   Latitude   Longitude
196            ALBION SQUARE  2.450000e+06 -41.273758  173.289393
390              ANHALT ROAD  2.435000e+06  51.480316   -0.166801
405          ANSDELL TERRACE  2.250000e+06  51.499890   -0.189103
422          APPLEGARTH ROAD  2.400000e+06  53.748654   -0.326670
855          BARONSMEAD ROAD  2.375000e+06  51.477315   -0.239457
981           BEAUCLERC ROAD  2.480000e+06  30.211452  -81.617981
1102         BELVEDERE DRIVE  2.340000e+06  38.072818  -78.458796
1215       BICKENHALL STREET  2.208500e+06  51.521201   -0.158908
1253       BIRCHLANDS AVENUE  2.217000e+06  51.448394   -0.160468
1553          BRAMPTON GROVE  2.456875e+06  51.589961   -0.318525
1632       BRIARDALE GARDENS  2.397132e+06  51.560175   -0.195431
1797                BROOKWAY  2.400000e+06  45.432185 -122.802812
1914            BURBAGE ROAD  2.445000e+06  52.538507   -1.353674
1980               BURY WALK  2.492500e+06  52.145529   -0.423593
2068         CALLCOTT STREET  2.375000e+06  51.508350   -0.198328
2129       CAMPDEN HILL ROAD  2.379653e+06  51.508111   -0.199667
2136            CAMPION ROAD  2.461000e+06  52.681078    0.965599
2158           CANNING PLACE  2.425000e+06  52.641027   -1.135223
2225           CARLISLE ROAD  2.200000e+06  43.399438  -79.973215
2230         CARLTON GARDENS  2.483500e+06 -37.801943  144.971970
2242           CARLYLE COURT  2.300000e+06  32.972701  -97.173392
2406          CHALCOT SQUARE  2.286679e+06  51.541088   -0.155928
2484            CHARLES LANE  2.414000e+06  40.257626  -75.093891
2562        CHELSEA CRESCENT  2.495000e+06  34.522443  -85.443891
2606     CHESTER CLOSE NORTH  2.450000e+06  51.529205   -0.145081
2638            CHEYNE COURT  2.250000e+06  52.035905    0.727612
2641              CHEYNE ROW  2.410000e+06  51.483717   -0.169603
2686           CHISWICK MALL  2.287500e+06  51.487994   -0.246605
2761               CITY ROAD  2.468340e+06  51.529697   -0.097763
2808        CLARENDON STREET  2.250000e+06  42.284961  -71.345534
...                      ...           ...        ...         ...
10935   RUSSELL GARDENS MEWS  2.300000e+06  51.499918   -0.211944
11182         SETTLES STREET  2.487500e+06  32.239424 -101.457123
11257         SHELDON AVENUE  2.349542e+06  42.020924  -93.654224
11491          SOUTH END ROW  2.470000e+06  51.498746   -0.189079
11568    SOUTHWOOD LAWN ROAD  2.350000e+06  51.574627   -0.146238
11571         SOVEREIGN PARK  2.500000e+06 -27.493713  153.210143
11788  ST MARGARETS CRESCENT  2.216500e+06  17.097514  -88.616115
11824       ST OSWALDS PLACE  2.250000e+06  51.487207   -0.118534
11844       ST PETERS SQUARE  2.468730e+06  41.902235   12.457357
11881       STAFFORD TERRACE  2.355000e+06  53.680501   -1.518338
12221       SUTHERLAND PLACE  2.456000e+06  51.871138   -0.422900
12283          SYDNEY STREET  2.240833e+06  44.265684  -76.551469
12425            THAMES BANK  2.400000e+06  52.085669   -0.243276
12487            THE HEXAGON  2.335000e+06  51.453882   -0.977834
12766        TREDEGAR SQUARE  2.436666e+06  51.527093   -0.032184
12828         TRINITY STREET  2.317500e+06  50.338316   -4.795430
12978   UPPER HAMPSTEAD WALK  2.500000e+06  51.558467   -0.177453
13240        WALPOLE GARDENS  2.303500e+06  51.439743   -0.341263
13242         WALPOLE STREET  2.242500e+06 -27.872584  151.271777
13316         WARWICK SQUARE  2.432273e+06  33.738563 -117.846506
13415            WELBECK WAY  2.267000e+06  52.558005   -0.262309
13427      WELLESLEY TERRACE  2.410000e+06  54.974313   -1.640202
13438      WELLINGTON STREET  2.293155e+06  45.423449  -75.698058
13570     WESTMORELAND PLACE  2.300000e+06  40.740224 -111.846323
13677       WHITFIELD STREET  2.451000e+06 -12.460726  130.839585
13722         WILFRED STREET  2.410538e+06  51.442178    0.372743
13748     WILLOW BRIDGE ROAD  2.425000e+06  53.640874   -1.200462
13768          WILSON STREET  2.257500e+06  42.286454  -71.400981
13796        WINCHENDON ROAD  2.350000e+06  42.645331  -71.954

In [69]:
address = 'London, UK'
geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinate of London : {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:2: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  from ipykernel import kernelapp as app


The coordinate of London : 51.5073219, -0.1276474.


In [70]:
map_london = folium.Map(location=[latitude, longitude], zoom_start=11) # Map creation
for lat, lng, price, street in zip(df['Latitude'], df['Longitude'], df['Avg_Price'], df['Street']):
    label = '{}, {}'.format(street, price)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_london)  
    
map_london

In [71]:
CLIENT_ID = 'QPCEJCW55ZLH4J3YHDFODAIGJAOZUHLGYPK0ORFA30T1RUA1' # Foursquare Client ID
CLIENT_SECRET = 'FVZUDXUVPDPXO2Q5K1F3LJP003YWLAZWHLORRNKVB0HV2GXP' # Foursquare Secret code
VERSION = '20181206'

### Step 4:  Data Modeling

We will dissect neighborhoods to suggest genuine bequests where home purchasers can make a land speculation. We will at that point prescribe productive settings as per pleasantries and fundamental offices encompassing such scenes for example primary schools, secondary schools, medical clinics and basic food item stores.After investigating the dataset and picking up bits of knowledge into it, we are prepared to utilize the bunching philosophy to dissect genuine bequests. We will utilize the k-implies grouping procedure as it is quick and proficient as far as computational expense, is exceptionally adaptable to represent transformations in land showcase in London and is precise.

In [72]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant info w.r.t venues
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Street', 
                  'Street Latitude', 
                  'Street Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [73]:
# Displaying new dataframe : location_venues
location_venues = getNearbyVenues(names=df['Street'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

ALBION SQUARE
ANHALT ROAD
ANSDELL TERRACE
APPLEGARTH ROAD
BARONSMEAD ROAD
BEAUCLERC ROAD
BELVEDERE DRIVE
BICKENHALL STREET
BIRCHLANDS AVENUE
BRAMPTON GROVE
BRIARDALE GARDENS
BROOKWAY
BURBAGE ROAD
BURY WALK
CALLCOTT STREET
CAMPDEN HILL ROAD
CAMPION ROAD
CANNING PLACE
CARLISLE ROAD
CARLTON GARDENS
CARLYLE COURT
CHALCOT SQUARE
CHARLES LANE
CHELSEA CRESCENT
CHESTER CLOSE NORTH
CHEYNE COURT
CHEYNE ROW
CHISWICK MALL
CITY ROAD
CLARENDON STREET
CLONCURRY STREET
COLBECK MEWS
COLLEGE CRESCENT
CORNWALL TERRACE MEWS
COURT LANE GARDENS
CRESCENT GROVE
DALEBURY ROAD
DEWHURST ROAD
DORIA ROAD
DOWNSHIRE HILL
DUCHESS WALK
ECCLESTON SQUARE MEWS
EGBERT STREET
EGERTON PLACE
ELM PARK ROAD
FLORAL STREET
FRANK DIXON WAY
FULTON MEWS
GERARD ROAD
GERRARD ROAD
GIRDLERS ROAD
GLOUCESTER CRESCENT
GORDON PLACE
GRAFTON SQUARE
GRAHAM TERRACE
HARMAN DRIVE
HARRIS STREET
HAVANNAH STREET
HAZLEWELL ROAD
HEREFORD MEWS
HERONDALE AVENUE
HIGHGATE HIGH STREET
HIGHWOOD HILL
HILLGATE PLACE
HOLLYCROFT AVENUE
HOLLYWOOD MEWS
HONEYWELL

In [74]:
location_venues

Street  Street Latitude  Street Longitude  \
0      ALBION SQUARE       -41.273758        173.289393   
1      ALBION SQUARE       -41.273758        173.289393   
2      ALBION SQUARE       -41.273758        173.289393   
3      ALBION SQUARE       -41.273758        173.289393   
4      ALBION SQUARE       -41.273758        173.289393   
5      ALBION SQUARE       -41.273758        173.289393   
6      ALBION SQUARE       -41.273758        173.289393   
7      ALBION SQUARE       -41.273758        173.289393   
8      ALBION SQUARE       -41.273758        173.289393   
9      ALBION SQUARE       -41.273758        173.289393   
10     ALBION SQUARE       -41.273758        173.289393   
11     ALBION SQUARE       -41.273758        173.289393   
12     ALBION SQUARE       -41.273758        173.289393   
13     ALBION SQUARE       -41.273758        173.289393   
14     ALBION SQUARE       -41.273758        173.289393   
15     ALBION SQUARE       -41.273758        173.289393   
16     ALBION SQUARE       -41.273758        173.289393   
17     ALBION SQUARE       -41.273758        173.289393   
18     ALBION SQUARE       -41.273758        173.289393   
19     ALBION SQUARE       -41.273758        173.289393   
20     ALBION SQUARE       -41.273758        173.289393   
21     ALBION SQUARE       -41.273758        173.289393   
22     ALBION SQUARE       -41.273758        173.289393   
23     ALBION SQUARE       -41.273758        173.289393   
24     ALBION SQUARE       -41.273758        173.289393   
25     ALBION SQUARE       -41.273758        173.289393   
26     ALBION SQUARE       -41.273758        173.289393   
27     ALBION SQUARE       -41.273758        173.289393   
28       ANHALT ROAD        51.480316         -0.166801   
29       ANHALT ROAD        51.480316         -0.166801   
...              ...              ...               ...   
4335  WILFRED STREET        51.442178          0.372743   
4336  WILFRED STREET        51.442178          0.372743   
4337  WILFRED STREET        51.442178          0.372743   
4338  WILFRED STREET        51.442178          0.372743   
4339  WILFRED STREET        51.442178          0.372743   
4340  WILFRED STREET        51.442178          0.372743   
4341  WILFRED STREET        51.442178          0.372743   
4342  WILFRED STREET        51.442178          0.372743   
4343  WILFRED STREET        51.442178          0.372743   
4344  WILFRED STREET        51.442178          0.372743   
4345  WILFRED STREET        51.442178          0.372743   
4346  WILFRED STREET        51.442178          0.372743   
4347  WILFRED STREET        51.442178          0.372743   
4348  WILFRED STREET        51.442178          0.372743   
4349  WILFRED STREET        51.442178          0.372743   
4350  WILFRED STREET        51.442178          0.372743   
4351  WILFRED STREET        51.442178          0.372743   
4352  WILFRED STREET        51.442178          0.372743   
4353  WILFRED STREET        51.442178          0.372743   
4354  WILFRED STREET        51.442178          0.372743   
4355  WILFRED STREET        51.442178          0.372743   
4356  WILFRED STREET        51.442178          0.372743   
4357  WILFRED STREET        51.442178          0.372743   
4358   WILSON STREET        42.286454        -71.400981   
4359   WILSON STREET        42.286454        -71.400981   
4360   WILSON STREET        42.286454        -71.400981   
4361    WINGATE ROAD        51.092557          1.179455   
4362    WINGATE ROAD        51.092557          1.179455   
4363    WINGATE ROAD        51.092557          1.179455   
4364    WINGATE ROAD        51.092557          1.179455   

                                      Venue  Venue Latitude  Venue Longitude  \
0                            The Free House      -41.273340       173.287364   
1                           The Indian Cafe      -41.273308       173.286530   
2                           Queen's Gardens      -41.273671       173.291383   
3                                  

In [75]:
location_venues.groupby('Street').count() # Count

Street Latitude  Street Longitude  Venue  \
Street                                                           
ALBION SQUARE                      28                28     28   
ANHALT ROAD                        17                17     17   
ANSDELL TERRACE                    43                43     43   
APPLEGARTH ROAD                     4                 4      4   
BARONSMEAD ROAD                    13                13     13   
BEAUCLERC ROAD                      5                 5      5   
BELVEDERE DRIVE                     3                 3      3   
BICKENHALL STREET                  59                59     59   
BIRCHLANDS AVENUE                  11                11     11   
BRAMPTON GROVE                      1                 1      1   
BRIARDALE GARDENS                   6                 6      6   
BROOKWAY                            1                 1      1   
BURBAGE ROAD                        4                 4      4   
BURY WALK                          15                15     15   
CALLCOTT STREET                    47                47     47   
CAMPDEN HILL ROAD                  46                46     46   
CAMPION ROAD                        4                 4      4   
CANNING PLACE                      12                12     12   
CARLTON GARDENS                    63                63     63   
CARLYLE COURT                       1                 1      1   
CHALCOT SQUARE                     54                54     54   
CHARLES LANE                        2                 2      2   
CHESTER CLOSE NORTH                22                22     22   
CHEYNE COURT                       14                14     14   
CHEYNE ROW                         60                60     60   
CHISWICK MALL                       7                 7      7   
CITY ROAD                          29                29     29   
CLARENDON STREET                   32                32     32   
CLONCURRY STREET                   31                31     31   
COLBECK MEWS                      100               100    100   
...                               ...               ...    ...   
RODERICK ROAD                       7                 7      7   
ROPEMAKERS FIELDS                  23                23     23   
ROYAL CRESCENT                     23                23     23   
RUSSELL GARDENS MEWS               26                26     26   
SETTLES STREET                      3                 3      3   
SHELDON AVENUE                     37                37     37   
SOUTH END ROW                      46                46     46   
SOUTHWOOD LAWN ROAD                34                34     34   
SOVEREIGN PARK                      3                 3      3   
ST OSWALDS PLACE                   41                41     41   
ST PETERS SQUARE                   78                78     78   
STAFFORD TERRACE                    7                 7      7   
SUTHERLAND PLACE                    2                 2      2   
SYDNEY STREET                       1                 1      1   
THAMES BANK                         4                 4      4   
THE HEXAGON                        38                38     38   
TREDEGAR SQUARE                    25                25     25   
TRINITY STREET                     18                18     18   
UPPER HAMPSTEAD WALK               60                60     60   
WALPOLE GARDENS                     9                 9      9   
WALPOLE STREET                      2                 2      2   
WARWICK SQUARE                      4                 4      4   
WELBECK WAY                         4                 4      4   
WELLESLEY TERRACE                  14                14     14   
WELLINGTON STREET                  87                87     87   
WESTMORELAND PLACE                 17                17     17   
WHITFIELD STREET                   42                42     42   
WILFRED STREET                     25                25     25

In [76]:
print('There are {} uniques categories.'.format(len(location_venues['Venue Category'].unique()))) 
# Extracting the number of unique categories.

There are 337 uniques categories.


In [77]:
location_venues.shape

(4365, 7)

In [78]:
venues_onehot = pd.get_dummies(location_venues[['Venue Category']], prefix="", prefix_sep="") # one hot encoding
venues_onehot['Street'] = location_venues['Street'] # adding the coln named street back to dataframe
fixed_columns = [venues_onehot.columns[-1]] + list(venues_onehot.columns[:-1]) # Assign street coln as 1st coln
venues_onehot = venues_onehot[fixed_columns]
venues_onehot.head()

Street  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  ALBION SQUARE                  0               0                  0   
1  ALBION SQUARE                  0               0                  0   
2  ALBION SQUARE                  0               0                  0   
3  ALBION SQUARE                  0               0                  0   
4  ALBION SQUARE                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Argentinian Restaurant  Art Gallery  ...  Volleyball Court  \
0                       0            0  ...                 0   
1                       0            0  ...                 0   
2                       0            0  ...                 0   
3                       0            0  ...                 0   
4                       0            0  ...                 0   

   Warehouse Store  Waterfront  Weight Loss Center  Windmill  Wine Bar  \
0                0           0                   0         0         0   
1                0           0                   0         0         0   
2                0           0                   0         0         0   
3                0           0                   0         0         0   
4                0           0                   0         0         0   

   Wine Shop  Women's Store  Yoga Studio  Zoo  
0          0              0            0    0  
1          0              0            0    0  
2          0              0            0    0  
3          0              0            0    0  
4          0              0            0    0  

[5 rows x 338 columns]

In [79]:
london_grouped = venues_onehot.groupby('Street').mean().reset_index()
london_grouped

Street  Accessories Store  Adult Boutique  \
0           ALBION SQUARE           0.000000             0.0   
1             ANHALT ROAD           0.000000             0.0   
2         ANSDELL TERRACE           0.000000             0.0   
3         APPLEGARTH ROAD           0.000000             0.0   
4         BARONSMEAD ROAD           0.000000             0.0   
5          BEAUCLERC ROAD           0.000000             0.0   
6         BELVEDERE DRIVE           0.000000             0.0   
7       BICKENHALL STREET           0.000000             0.0   
8       BIRCHLANDS AVENUE           0.000000             0.0   
9          BRAMPTON GROVE           0.000000             0.0   
10      BRIARDALE GARDENS           0.000000             0.0   
11               BROOKWAY           0.000000             0.0   
12           BURBAGE ROAD           0.000000             0.0   
13              BURY WALK           0.000000             0.0   
14        CALLCOTT STREET           0.000000             0.0   
15      CAMPDEN HILL ROAD           0.000000             0.0   
16           CAMPION ROAD           0.000000             0.0   
17          CANNING PLACE           0.000000             0.0   
18        CARLTON GARDENS           0.000000             0.0   
19          CARLYLE COURT           0.000000             0.0   
20         CHALCOT SQUARE           0.000000             0.0   
21           CHARLES LANE           0.000000             0.0   
22    CHESTER CLOSE NORTH           0.000000             0.0   
23           CHEYNE COURT           0.000000             0.0   
24             CHEYNE ROW           0.016667             0.0   
25          CHISWICK MALL           0.000000             0.0   
26              CITY ROAD           0.000000             0.0   
27       CLARENDON STREET           0.000000             0.0   
28       CLONCURRY STREET           0.000000             0.0   
29           COLBECK MEWS           0.000000             0.0   
..                    ...                ...             ...   
119         RODERICK ROAD           0.000000             0.0   
120     ROPEMAKERS FIELDS           0.000000             0.0   
121        ROYAL CRESCENT           0.000000             0.0   
122  RUSSELL GARDENS MEWS           0.000000             0.0   
123        SETTLES STREET           0.000000             0.0   
124        SHELDON AVENUE           0.000000             0.0   
125         SOUTH END ROW           0.000000             0.0   
126   SOUTHWOOD LAWN ROAD           0.000000             0.0   
127        SOVEREIGN PARK           0.000000             0.0   
128      ST OSWALDS PLACE           0.000000             0.0   
129      ST PETERS SQUARE           0.000000             0.0   
130      STAFFORD TERRACE           0.000000             0.0   
131      SUTHERLAND PLACE           0.000000             0.0   
132         SYDNEY STREET           0.000000             0.0   
133           THAMES BANK           0.000000             0.0   
134           THE HEXAGON           0.000000             0.0   
135       TREDEGAR SQUARE           0.000000             0.0   
136        TRINITY STREET           0.000000             0.0   
137  UPPER HAMPSTEAD WALK           0.000000             0.0   
138       WALPOLE GARDENS           0.000000             0.0   
139        WALPOLE STREET           0.000000             0.0   
140        WARWICK SQUARE           0.000000             0.0   
141           WELBECK WAY           0.000000             0.0   
142     WELLESLEY TERRACE           0.000000             0.0   
143     WELLINGTON STREET           0.000000             0.0   
144    WESTMORELAND PLACE           0.000000             0.0   
145      WHITFIELD STREET           0.000000             0.0   
146        WILFRED STREET           0.000000             0.0   
147         WILSON STREET           0.000000             0.0   
148          WINGATE ROAD           0.000000             0.0   

     Afghan Restaurant  African Restaurant  Americ

In [80]:
london_grouped.shape

(149, 338)

In [81]:
# Top 5 landmarks/venues where there are profitable real estate investments
num_top_venues = 5

for hood in london_grouped['Street']:
    print("----"+hood+"----")
    temp = london_grouped[london_grouped['Street'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----ALBION SQUARE----
               venue  freq
0               Café  0.21
1                Pub  0.07
2         Restaurant  0.07
3  Indian Restaurant  0.07
4                Bar  0.07


----ANHALT ROAD----
                  venue  freq
0                   Pub  0.24
1         Grocery Store  0.12
2  Gym / Fitness Center  0.06
3   Japanese Restaurant  0.06
4                 Plaza  0.06


----ANSDELL TERRACE----
                venue  freq
0          Restaurant  0.09
1           Juice Bar  0.07
2  Italian Restaurant  0.07
3               Hotel  0.07
4      Clothing Store  0.07


----APPLEGARTH ROAD----
          venue  freq
0           Pub  0.25
1     Nightclub  0.25
2        Casino  0.25
3           Bar  0.25
4  Noodle House  0.00


----BARONSMEAD ROAD----
               venue  freq
0    Nature Preserve  0.08
1  Food & Drink Shop  0.08
2      Movie Theater  0.08
3        Coffee Shop  0.08
4               Park  0.08


----BEAUCLERC ROAD----
                 venue  freq
0                  S

In [82]:
# Returning the most common landmarks/venues nearby real estate investments
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10
indicators = ['st', 'nd', 'rd']
columns = ['Street']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

In [83]:
# create a new dataframe
venues_sorted = pd.DataFrame(columns=columns)
venues_sorted['Street'] = london_grouped['Street']

for ind in np.arange(london_grouped.shape[0]):
    venues_sorted.iloc[ind, 1:] = return_most_common_venues(london_grouped.iloc[ind, :], num_top_venues)

In [84]:
venues_sorted.head()

Street 1st Most Common Venue 2nd Most Common Venue  \
0    ALBION SQUARE                  Café           Coffee Shop   
1      ANHALT ROAD                   Pub         Grocery Store   
2  ANSDELL TERRACE            Restaurant                 Hotel   
3  APPLEGARTH ROAD                   Pub                   Bar   
4  BARONSMEAD ROAD         Movie Theater           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Restaurant                   Bar                   Pub   
1          Cocktail Bar           Art Gallery  Gym / Fitness Center   
2             Juice Bar        Clothing Store    Italian Restaurant   
3             Nightclub                Casino       Coworking Space   
4             Bookstore       Nature Preserve        Farmers Market   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Indian Restaurant           Beer Garden                Museum   
1                 Diner                Garden                  Pier   
2                Bakery                   Pub                Garden   
3   Filipino Restaurant           Event Space               Exhibit   
4       Thai Restaurant        Breakfast Spot                  Park   

  9th Most Common Venue 10th Most Common Venue  
0     Fish & Chips Shop            Art Gallery  
1     French Restaurant     English Restaurant  
2     Indian Restaurant     Chinese Restaurant  
3               Factory     Falafel Restaurant  
4            Restaurant      Food & Drink Shop

In [85]:
venues_sorted.shape

(149, 11)

In [86]:
london_grouped.shape

(149, 338)

In [87]:
london_grouped=df

In [88]:
# Clutering the properties in 5 Clusters
kclusters = 5
london_grouped_clustering = london_grouped.drop('Street', 1)
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(london_grouped_clustering)# k-means clutering
kmeans.labels_[0:50]

array([1, 3, 0, 3, 2, 1, 2, 0, 0, 1, 3, 3, 3, 1, 2, 2, 1, 3, 0, 1, 4, 4,
       3, 1, 1, 0, 3, 4, 1, 0, 3, 2, 3, 2, 2, 4, 3, 3, 2, 0, 1, 2, 4, 0,
       4, 0, 0, 4, 0, 0], dtype=int32)

In [89]:
london_grouped_clustering=df # london_grouped_clustering to include dataframe
london_grouped_clustering.head()

Street  Avg_Price   Latitude   Longitude
196    ALBION SQUARE  2450000.0 -41.273758  173.289393
390      ANHALT ROAD  2435000.0  51.480316   -0.166801
405  ANSDELL TERRACE  2250000.0  51.499890   -0.189103
422  APPLEGARTH ROAD  2400000.0  53.748654   -0.326670
855  BARONSMEAD ROAD  2375000.0  51.477315   -0.239457

In [90]:
london_grouped_clustering.shape

(162, 4)

In [91]:
df.shape

(162, 4)

In [92]:
london_grouped_clustering.dtypes
df.dtypes

Street        object
Avg_Price    float64
Latitude     float64
Longitude    float64
dtype: object

In [93]:
df.dtypes

Street        object
Avg_Price    float64
Latitude     float64
Longitude    float64
dtype: object

In [94]:
london_grouped_clustering['Cluster Labels'] = kmeans.labels_ # Clustering labels
london_grouped_clustering = london_grouped_clustering.join(venues_sorted.set_index('Street'), on='Street') # merging london_grouped with london_data 
london_grouped_clustering.head(30) 

Street     Avg_Price   Latitude   Longitude  \
196         ALBION SQUARE  2.450000e+06 -41.273758  173.289393   
390           ANHALT ROAD  2.435000e+06  51.480316   -0.166801   
405       ANSDELL TERRACE  2.250000e+06  51.499890   -0.189103   
422       APPLEGARTH ROAD  2.400000e+06  53.748654   -0.326670   
855       BARONSMEAD ROAD  2.375000e+06  51.477315   -0.239457   
981        BEAUCLERC ROAD  2.480000e+06  30.211452  -81.617981   
1102      BELVEDERE DRIVE  2.340000e+06  38.072818  -78.458796   
1215    BICKENHALL STREET  2.208500e+06  51.521201   -0.158908   
1253    BIRCHLANDS AVENUE  2.217000e+06  51.448394   -0.160468   
1553       BRAMPTON GROVE  2.456875e+06  51.589961   -0.318525   
1632    BRIARDALE GARDENS  2.397132e+06  51.560175   -0.195431   
1797             BROOKWAY  2.400000e+06  45.432185 -122.802812   
1914         BURBAGE ROAD  2.445000e+06  52.538507   -1.353674   
1980            BURY WALK  2.492500e+06  52.145529   -0.423593   
2068      CALLCOTT STREET  2.375000e+06  51.508350   -0.198328   
2129    CAMPDEN HILL ROAD  2.379653e+06  51.508111   -0.199667   
2136         CAMPION ROAD  2.461000e+06  52.681078    0.965599   
2158        CANNING PLACE  2.425000e+06  52.641027   -1.135223   
2225        CARLISLE ROAD  2.200000e+06  43.399438  -79.973215   
2230      CARLTON GARDENS  2.483500e+06 -37.801943  144.971970   
2242        CARLYLE COURT  2.300000e+06  32.972701  -97.173392   
2406       CHALCOT SQUARE  2.286679e+06  51.541088   -0.155928   
2484         CHARLES LANE  2.414000e+06  40.257626  -75.093891   
2562     CHELSEA CRESCENT  2.495000e+06  34.522443  -85.443891   
2606  CHESTER CLOSE NORTH  2.450000e+06  51.529205   -0.145081   
2638         CHEYNE COURT  2.250000e+06  52.035905    0.727612   
2641           CHEYNE ROW  2.410000e+06  51.483717   -0.169603   
2686        CHISWICK MALL  2.287500e+06  51.487994   -0.246605   
2761            CITY ROAD  2.468340e+06  51.529697   -0.097763   
2808     CLARENDON STREET  2.250000e+06  42.284961  -71.345534   

      Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
196                1                  Café           Coffee Shop   
390                3                   Pub         Grocery Store   
405                0            Restaurant                 Hotel   
422                3                   Pub                   Bar   
855                2         Movie Theater           Coffee Shop   
981                1                   Spa           Pizza Place   
1102               2                  Pool            Playground   
1215               0                  Café           Pizza Place   
1253               0     French Restaurant                   Pub   
1553               1          Home Service                   Zoo   
1632               3  Gym / Fitness Center    Italian Restaurant   
1797               3           Art Gallery                   Zoo   
1914               3                   Bar         Grocery Store   
1980               1           Supermarket    English Restaurant   
2068               2                   Pub                 Hotel   
2129               2                   Pub         Grocery Store   
2136               1          Soccer Field                   Spa   
2158               3        Clothing Store     Convenience Store   
2225               0                   NaN                   NaN   
2230               1    Italian Restaurant                  Café   
2242               4                  Farm                   Zoo   
2406               4                   Pub                  Café   
2484               3   American Restaurant    Photography Studio   
2562               1                   NaN                   NaN   
2606               1                Garden  Gym / Fitness Center   
2638               0            Restaurant              Pharmacy   
2641               3                  Café    Italian Restaurant   
2686               4                   Pub            Art Museum   


In [95]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11) # Map creation and setting color scheme
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]
markers_colors = []
for lat, lon, poi, cluster in zip(london_grouped_clustering['Latitude'], london_grouped_clustering['Longitude'], london_grouped_clustering['Street'], london_grouped_clustering['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [96]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 0, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
405   2250000.0            Restaurant                 Hotel   
1215  2208500.0                  Café           Pizza Place   
1253  2217000.0     French Restaurant                   Pub   
2225  2200000.0                   NaN                   NaN   
2638  2250000.0            Restaurant              Pharmacy   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
405              Juice Bar        Clothing Store    Italian Restaurant   
1215                 Hotel            Restaurant                Bakery   
1253                Bakery           Coffee Shop  Gym / Fitness Center   
2225                   NaN                   NaN                   NaN   
2638    Seafood Restaurant                  Bank                Bakery   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
405                 Bakery                   Pub                Garden   
1215         Movie Theater             Gastropub                Garden   
1253    Chinese Restaurant         Train Station                  Lake   
2225                   NaN                   NaN                   NaN   
2638         Grocery Store           Supermarket                  Café   

     9th Most Common Venue 10th Most Common Venue  
405      Indian Restaurant     Chinese Restaurant  
1215                   Bar     Italian Restaurant  
1253               Brewery           Dance Studio  
2225                   NaN                    NaN  
2638  Outdoor Supply Store            Coffee Shop

In [97]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 1, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
196   2450000.0                  Café           Coffee Shop   
981   2480000.0                   Spa           Pizza Place   
1553  2456875.0          Home Service                   Zoo   
1980  2492500.0           Supermarket    English Restaurant   
2136  2461000.0          Soccer Field                   Spa   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
196             Restaurant                   Bar                   Pub   
981        Automotive Shop       Harbor / Marina                   Zoo   
1553  Fast Food Restaurant    English Restaurant  Ethiopian Restaurant   
1980                   Gym   Rental Car Location                  Café   
2136              Windmill           Bus Station  Fast Food Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
196      Indian Restaurant           Beer Garden                Museum   
981                   Farm    English Restaurant  Ethiopian Restaurant   
1553           Event Space               Exhibit               Factory   
1980           Coffee Shop  Fast Food Restaurant        Hardware Store   
2136  Ethiopian Restaurant           Event Space               Exhibit   

     9th Most Common Venue 10th Most Common Venue  
196      Fish & Chips Shop            Art Gallery  
981            Event Space                Exhibit  
1553    Falafel Restaurant                   Farm  
1980   American Restaurant                    Pub  
2136               Factory     Falafel Restaurant

In [98]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 2, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
855   2375000.0         Movie Theater           Coffee Shop   
1102  2340000.0                  Pool            Playground   
2068  2375000.0                   Pub                 Hotel   
2129  2379652.7                   Pub         Grocery Store   
2944  2367500.0                 Hotel                   Pub   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
855              Bookstore       Nature Preserve        Farmers Market   
1102         Auto Workshop                   Zoo        Farmers Market   
2068         Grocery Store                  Park        Ice Cream Shop   
2129     Indian Restaurant                Bakery                  Park   
2944                Garden                  Café           Coffee Shop   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
855        Thai Restaurant        Breakfast Spot                  Park   
1102    English Restaurant  Ethiopian Restaurant           Event Space   
2068     Indian Restaurant           Pizza Place           Yoga Studio   
2129           Coffee Shop           Yoga Studio                 Hotel   
2944    Italian Restaurant    Chinese Restaurant                   Bar   

         9th Most Common Venue                    10th Most Common Venue  
855                 Restaurant                         Food & Drink Shop  
1102                   Exhibit                                   Factory  
2068                    Hostel                                 Bookstore  
2129            Ice Cream Shop                               Pizza Place  
2944  Mediterranean Restaurant  Residential Building (Apartment / Condo)

In [99]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 3, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
390   2435000.0                   Pub         Grocery Store   
422   2400000.0                   Pub                   Bar   
1632  2397132.0  Gym / Fitness Center    Italian Restaurant   
1797  2400000.0           Art Gallery                   Zoo   
1914  2445000.0                   Bar         Grocery Store   

     3rd Most Common Venue    4th Most Common Venue 5th Most Common Venue  \
390           Cocktail Bar              Art Gallery  Gym / Fitness Center   
422              Nightclub                   Casino       Coworking Space   
1632         Grocery Store  Health & Beauty Service           Coffee Shop   
1797   Filipino Restaurant     Ethiopian Restaurant           Event Space   
1914          Dance Studio       Athletics & Sports  Fast Food Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
390                  Diner                Garden                  Pier   
422    Filipino Restaurant           Event Space               Exhibit   
1632        Breakfast Spot        Cricket Ground  Fast Food Restaurant   
1797               Exhibit               Factory    Falafel Restaurant   
1914  Ethiopian Restaurant           Event Space               Exhibit   

     9th Most Common Venue 10th Most Common Venue  
390      French Restaurant     English Restaurant  
422                Factory     Falafel Restaurant  
1632  Ethiopian Restaurant            Event Space  
1797                  Farm         Farmers Market  
1914               Factory     Falafel Restaurant

In [100]:
london_grouped_clustering.loc[london_grouped_clustering['Cluster Labels'] == 4, london_grouped_clustering.columns[[1] + list(range(5, london_grouped_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
2242  2.300000e+06                  Farm                   Zoo   
2406  2.286679e+06                   Pub                  Café   
2686  2.287500e+06                   Pub            Art Museum   
3377  2.298000e+06                 Hotel                   Zoo   
4285  2.265000e+06                   Pub         Shopping Mall   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2242  Fast Food Restaurant    English Restaurant  Ethiopian Restaurant   
2406    Italian Restaurant                   Bar           Coffee Shop   
2686               Brewery             Gift Shop  Gym / Fitness Center   
3377  Fast Food Restaurant    English Restaurant  Ethiopian Restaurant   
4285                  Farm     Electronics Store    English Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2242           Event Space               Exhibit               Factory   
2406     Convenience Store                  Park     French Restaurant   
2686              Creperie   Filipino Restaurant               Exhibit   
3377           Event Space               Exhibit               Factory   
4285  Ethiopian Restaurant           Event Space               Exhibit   

     9th Most Common Venue 10th Most Common Venue  
2242    Falafel Restaurant         Farmers Market  
2406        Breakfast Spot     Belgian Restaurant  
2686               Factory           Costume Shop  
3377    Falafel Restaurant                   Farm  
4285               Factory     Falafel Restaurant

## Results section

As a matter of first importance, despite the fact that the London Housing Market might be stuck, it is as yet an "ever-green" for business undertakings. We may talk about our outcomes under two primary viewpoints. To begin with, we may look at them as per neighborhoods/London zones. It is fascinating to take note of that, albeit West London (Notting Hill, Kensington, Chelsea, Marylebone) and North-West London (Hampsted) may be viewed as profoundly beneficial settings to buy a land as per civilities and basic offices encompassing such scenes for example primary schools, secondary schools, clinics and supermarkets, South-West London (Wandsworth, Balham) and North-West London (Isliington) are emerging as next future world class settings with a wide scope of conveniences and offices. In like manner, one may focus under-valued genuine bequests in these territories of London so as to make a business issue. Secondly, we may investigate our outcomes as per the five groups we have delivered. Despite the fact that, all bunches could commend an ideal scope of offices and pleasantries, we have discovered two primary examples. The principal design we are alluding to, for example Groups 0, 2 and 4, may target home purchasers inclined to live in 'green' regions with parks, waterfronts. Rather, the second example we are alluding to, for example Bunches 1 and 3, may target people who love bars, theaters and soccer.

## Concluding the Report

To summarize, as per Bloomberg News, the London Housing Market is stuck. It is presently confronting various headwinds, including the possibility of higher duties and an admonition from the Bank of England that U.K. home estimations could fall as much as 30 percent in case of a tumultuous exit from the European Union. In this situation, it is dire to embrace AI devices so as to help homebuyers customer base in London to settle on shrewd and viable choices. Thus, the business issue we were presenting was: how might we offer help to homebuyers customer base in to buy a reasonable land in London in this dubious monetary and money related situation? To take care of this business issue, we grouped London neighborhoods so as to suggest settings and the present normal cost of land where homebuyers can make a land venture. We prescribed beneficial scenes as per luxuries and basic offices encompassing such settings for example primary schools, secondary schools, medical clinics and supermarkets. To start with, we accumulated information on London properties and the relative cost paid information were extricated from the HM Land Registry (http://landregistry.data.gov.uk/). Also, to investigate and target prescribed areas across various settings as indicated by the nearness of luxuries and fundamental offices, we got to information through FourSquare API interface and masterminded them as an information outline for representation. By blending information on London properties and the relative cost paid information from the HM Land Registry and information on conveniences and fundamental offices encompassing such properties from FourSquare API interface, we had the option to suggest beneficial land ventures. Second, The Methodology segment contained four phases: 1. Gather Inspection Data; 2. Investigate and Understand Data; 3. Information readiness and preprocessing; 4. Demonstrating. Specifically, in the displaying area, we utilized the k-implies grouping strategy as it is quick and proficient as far as computational expense, is profoundly adaptable to represent changes in land showcase in London and is precise. At long last, we made the inference that despite the fact that the London Housing Market might be stuck, it is as yet an "ever-green" for business undertakings. We talked about our outcomes under two fundamental points of view. To begin with, we analyzed them as per neighborhoods/London zones. albeit West London (Notting Hill, Kensington, Chelsea, Marylebone) and North-West London (Hampsted) may be viewed as profoundly gainful scenes to buy a land as per comforts and fundamental offices encompassing such settings for example grade schools, secondary schools, medical clinics and supermarkets, South-West London (Wandsworth, Balham) and North-West London (Isliington) are emerging as next future tip top scenes with a wide scope of courtesies and offices. Likewise, one may focus under-estimated genuine homes in these territories of London so as to make a business undertaking. Second, we examined our outcomes as indicated by the five groups we created. While Clusters 0, 2 and 4 may target home purchasers inclined to live in 'green' zones with parks, waterfronts, Clusters 1 and 3 may target people who love bars, theaters and soccer.